src/Simulation/CSharpGeneration/SimulationCode.fs

```diff
@@ -1052,7 +1051,7 @@ module SimulationCode =
             let rec argsToVars = function
                 | QsTupleItem one -> [one.VariableName |> name]
                 | QsTuple many -> many |> Seq.map argsToVars |> List.concat
-            let callExp = (``ident`` "Gate") <.> (``ident`` specCall, argsToVars args)
+            let callExp = ``((`` (``cast`` ("IIntrinsic" + (userDefinedName None op.FullName.Name)) (``ident`` "Impl")) ``))`` <.> (``ident`` specCall, argsToVars args)
```

Comment:
@cgranade With this change, a simulator that tries to invoke a gate that it doesn't implement the interface for will get an invalid cast error instead of a null dereference. Is that sufficient to address #513 ?